### PR TITLE
Run the agent process as non-root user

### PR DESCRIPTION
**Pull request by akkie**
In kubernetes clusters is common to enable pod security policies, that disallow containers that run the main process as root, because hackers can gain root access to the Docker host by hacking the application running inside the container.

With this PR the blackfire agent runs as an unprivileged user.

### CODE REVIEW

File: Dockerfile
Comment (lyrixx) [inline Dockerfile]: I think the use could be a `system` user, and this use does not need to have a `$HOME` folder.
so:
```bash
adduser --system -D -H  blackfire
```

File: Dockerfile
Comment (lyrixx) [inline Dockerfile]: Why did you change the whole directory? There is no need for that
(Same comment for the next line)

File: Dockerfile
Comment (akkie) [reply Dockerfile]: I've opened a [question on SO](https://stackoverflow.com/questions/57458922/run-process-in-docker-container-as-system-or-normal-user) regarding this issue. It was put as on hold. So I'm not sure I'll get an answer.

If I should answer it, I would say that a UID in the container should't match a UID on the host system. The issue is that the UID allocation on the host system isn't really predictable because many systems use different ranges for UID allocations. It seems that on most modern Linux system, users will be randomly allocated in the range between 100 - 999 and normal users will be allocated in an increasing order, on most modern Linux system in the range from 1000 - 60000. So wouldn't it make sense to choose a UID of a normal user in the range between 5000 - 6000 or even higher? Choosing a system user would increase the risk to match an existing system user on the host system. But if I choose a very high UID in the normal user range, the change to match a user on a host system would be very low.

I think there is no general guidance. So let me know what you prefer and I will update the PR.

File: Dockerfile
Comment (akkie) [reply Dockerfile]: From https://unix.stackexchange.com/questions/8656/usr-bin-vs-usr-local-bin-on-linux

> /usr/local/bin is for normal user programs not managed by the distribution package manager, e.g. locally compiled packages. You should not install them into /usr/bin because future distribution upgrades may modify or delete them without warning.

For the sake of consistency, I thought I'll follow this advice. But the same as in my other answer. Let me know what you prefer and I will update the PR.

File: Dockerfile
Comment (lyrixx) [reply Dockerfile]: I tested with postgres, redis, rabbitmq, mysql, they all use a system user with ID 999

**Comment by akkie** [comment]
I've updated the PR to run the agent as system user with gid 999 and uid 999

File: Dockerfile
Comment (lyrixx) [inline Dockerfile]: This does not work, the group 999 already exists on alpine

Comment (akkie) [reply Dockerfile]: OK, I've removed it

File: Dockerfile
Comment (lyrixx) [inline Dockerfile]: You forgot the `-H` option to not create the `/home/blackfire` dir

Comment (akkie) [reply Dockerfile]: Sorry, fixed

**Comment by akkie** [comment]
Is the PR now OK? Let me know if there's anything else to change!

File: Dockerfile
Comment (iamluc) [reply Dockerfile]: I think `chown` and `chmod` commands are not needed at all.

Comment (akkie) [reply Dockerfile]: I thought it would be good that the user who runs the main process owns the files and that the permissions are explicitly set, even if the executable bit is set in git.

But, your'e right. It's not really needed. I can remove it if desired.

Comment (iamluc) [reply Dockerfile]: I'm NOT a security expert, but it think a security breach in the agent would allow the hacker to change the binary.

Also, I checked with Debian and Alpine, all binaries are owned by root (with some rare exceptions).

IMHO, we should do the same.

Comment (akkie) [reply Dockerfile]: Alright, done :+1:

**Review by iamluc — APPROVED** [review]
:gem:  
Thanks a lot.

Before merging, could you squash the commits?

**Comment by akkie** [comment]
Sure, done. You should consider to enable "Allow squash merging" and "Allow rebase merging" in the repo settings. So you could squash the commits directly in GitHub.

**Comment by iamluc** [comment]
@akkie in fact, I cannot merge this PR into this repository. Instead I must copy your changes in an internal script of Blackfire, so you will loose the ownership of the patch.

I hope this is OK for you?
Sorry about that.

**Comment by akkie** [comment]
Sure, no problem :+1:

**Review by lyrixx — APPROVED** [review]

**Comment by iamluc** [comment]
The changes will be available with the next stable release.

Thanks @akkie (and @lyrixx for the review ;-) )

**Pull request closed without merging**
